tests/experimental/test_experimental_utils.py

```diff
@@ -0,0 +1,508 @@
+            self.mixin.check_matrix(tensor_1d)
+
+    def test_check_pair_valid(self, tensor_2d, tensor_2d_pair):
+        """
+        Test check_pair with valid tensors.
+        """
+        x_checked, y_checked = self.mixin.check_pair(tensor_2d, tensor_2d_pair)
+
+        assert torch.equal(x_checked, tensor_2d)
+        assert torch.equal(y_checked, tensor_2d_pair)
+
+    def test_check_pair_invalid(self, tensor_2d, tensor_2d_mismatch):
+        """
+        Test check_pair with tensors of mismatched rows.
+        """
+        with pytest.raises(
+            ValueError, match="X and Y must have the same number of rows"
+        ):
+            self.mixin.check_pair(tensor_2d, tensor_2d_mismatch)
+
+    def test_check_covariance_valid(self, sigma_full, tensor_2d, tensor_1d):
+        """
+        Test check_covariance with valid covariance matrices.
+
+        Explanation:
+        - `sigma_full`: Represents a full covariance matrix for each sample in `y`.
+          It has a shape of (n_samples, n_features, n_features), where each sample
+          has a full covariance matrix.
+        - `sigma_diag`: Represents a diagonal covariance matrix for each sample in `y`.
+          It has a shape of (n_samples, n_features), where each row corresponds to
+          the diagonal elements of the covariance matrix for a sample.
+        - `sigma_scalar`: Represents a scalar covariance value for each sample in `y`.
+          It has a shape of (n_samples,), where each value corresponds to the same
+          scalar covariance for all features of a sample.
+
+        The test ensures that the `check_covariance` method correctly validates and
+        returns these different types of covariance matrices without modification.
+        """
+        y = torch.tensor([[1.0, 2.0], [3.0, 4.0]])
+
+        # Diagonal covariance matrix: shape (n_samples, n_features)
+        sigma_diag = tensor_2d
+
+        # Scalar covariance: shape (n_samples,)
+        sigma_scalar = tensor_1d
+
+        # Assert that the method returns the same full covariance matrix
+        assert torch.equal(self.mixin.check_covariance(y, sigma_full), sigma_full)
+
+        # Assert that the method returns the same diagonal covariance matrix
+        assert torch.equal(self.mixin.check_covariance(y, sigma_diag), sigma_diag)
+
+        # Assert that the method returns the same scalar covariance values
+        assert torch.equal(self.mixin.check_covariance(y, sigma_scalar), sigma_scalar)
+
+    def test_check_covariance_invalid(self, tensor_2d):
+        """
+        Test check_covariance with an invalid covariance matrix.
+        """
+        sigma = torch.tensor([[1.0, 2.0]])
+        with pytest.raises(ValueError, match="Invalid covariance matrix shape"):
+            self.mixin.check_covariance(tensor_2d, sigma)
+
+    def test_trace(self, sigma_full, tensor_2d, tensor_1d):
+        """
+        Test trace computation for covariance matrices.
```

Comment:
@cisprague I've had a go at adding tests for most of the Validation functions, but I wasn't sure on the `trace` and `logdet` functions - it may be worth you having a review of how I've set all of these up

Reply:
Leave for future PR